util/filesort/filesort.go

```
@@ -0,0 +1,465 @@
+	)
+	n, err = fs.fds[index].Read(head)
+	if err == io.EOF {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != 8 {
+		return nil, errors.New("incorrect header")
+	}
+	rowSize := int(binary.BigEndian.Uint64(head))
+
+	rowBytes := make([]byte, rowSize)
+
+	n, err = fs.fds[index].Read(rowBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != rowSize {
+		return nil, errors.New("incorrect row")
+	}
+
+	dcod, err = codec.Decode(rowBytes, fs.keySize+fs.valSize+1)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &comparableRow{
+		key:    dcod[:fs.keySize],
+		val:    dcod[fs.keySize : fs.keySize+fs.valSize],
+		handle: dcod[fs.keySize+fs.valSize:][0].GetInt64(),
+	}, nil
+}
+
+func (fs *FileSorter) openAllFiles() error {
+	for _, fname := range fs.files {
+		fd, err := os.Open(fname)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		fs.fds = append(fs.fds, fd)
+	}
+	return nil
+}
+
+func (fs *FileSorter) closeAllFiles() error {
+	for _, fd := range fs.fds {
+		err := fd.Close()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	err := os.RemoveAll(fs.tmpDir)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Close terminates the output process and discards all remaining data
+func (fs *FileSorter) Close() error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
+	}
+	fs.buf = fs.buf[:0]
+	fs.closed = true
+	return fs.closeAllFiles()
+}
+
+// Output gets the next sorted row.
+func (fs *FileSorter) Output() (key []types.Datum, val []types.Datum, handle int64, err error) {
+	if fs.closed {
+		return nil, nil, 0, errors.New("FileSorter has been closed")
+	}
+	if !fs.fetched {
+		if len(fs.buf) > 0 {
+			err = fs.flushMemory()
+			if err != nil {
+				return nil, nil, 0, errors.Trace(err)
+			}
+		}
+
+		heap.Init(fs.rowHeap)
+		if fs.rowHeap.err != nil {
+			return nil, nil, 0, errors.Trace(fs.rowHeap.err)
+		}
+
+		fs.openAllFiles()
```

error is not checked.